linera-core/src/client/mod.rs

```diff
@@ -2496,33 +2393,91 @@ where
             }
         }
 
-        // The block we want to propose is either the highest validated, or our pending one.
-        let Some(block) = manager
-            .highest_validated_block()
-            .cloned()
-            .or_else(|| self.state().pending_block().clone())
-        else {
-            return Ok(ClientOutcome::Committed(None)); // Nothing to propose.
+        // Otherwise we have to re-propose the highest validated block, if there is one.
+        let executed_block = if let Some(certificate) = &info.manager.requested_locked {
+            certificate.executed_block().clone()
+        } else {
+            let block = if let Some(proposal) = info
+                .manager
+                .requested_proposed
+                .as_ref()
+                .filter(|proposal| proposal.content.round.is_fast())
+            {
+                // The fast block counts as "locked", too. If there is one, re-propose that.
+                proposal.content.block.clone()
+            } else if let Some(block) = self.state().pending_block() {
+                block.clone() // Otherwise we are free to propose our own pending block.
+            } else {
+                return Ok(ClientOutcome::Committed(None)); // Nothing to do.
+            };
+            self.stage_block_execution(block).await?.0
         };
 
-        let round = match Self::round_for_new_proposal(&info, &identity, Some(&block))? {
-            Either::Left(round) => round,
-            Either::Right(timeout) => return Ok(ClientOutcome::WaitForTimeout(timeout)),
+        let round = match Self::round_for_new_proposal(
+            &info.manager,
+            &identity,
+            Some(&executed_block.block),
```

Comment:
If the block here is from `info.manager.requested_locked` then the `round` here will be `certificate.round + 1`, right? b/c if `requested_locked` was `Some(_)` and its `round` was `info.manager.current_round` then we would have called `finalized_current_round` 🤔

Reply:
…or greater, yes.

If it was `current_round`, we'd have tried to finalize it, yes. If it isn't, that means the current round is strictly greater. It could be `+ 2` or `+ 3` or more as well, though, if all the rounds in between have simply timed out.

Reply:
But we wouldn't have known that, right? b/c we don't update the `ChainInfo` between these two places in the code.

Reply:
No, even with the same chain info: The locked block is not necessarily in one of the last two rounds. In this one chain info it can happen that `current_round` is `SingleLeader(7)` but the locked block is in `SingleLeader(3)`.